src/providers-registry/providers-registry-contract.ts

```diff
@@ -0,0 +1,270 @@
+      throw new ContractError("New contract admins not defined");
+    }
+    checkIsContractAdmin(caller);
+    state.contractAdmins.push(...data.admins);
+
+    return {state }
+  }
+
+
+  /* STATE READ_ONLY ACTIONS */
+  // admin-only: returns last uploaded providers-registry - ie. last element in the manifests array
+  if (input.function === "provider-data") {
+    const data = input.data as GetProviderData;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+
+    return {
+      result: {
+        provider: providers[data.providerId]
+      }
+    }
+  }
+
+  if (input.function === "active-manifest") {
+    const data = input.data as GetProviderManifest;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+
+    const manifests = getManifestsForCaller(caller);
+    trace("Block height: ", blockHeight);
+    const manifest = manifests.slice().reverse().find((manifest) => {
+      trace("Checking manifest: ", manifest);
+      trace("Checking lockedBlocks: ", manifest.lockedHours * BLOCKS_IN_HOUR);
+      return manifest.uploadBlockHeight + (manifest.lockedHours * BLOCKS_IN_HOUR) <= blockHeight;
+    }) || manifests[0];
+
+    trace("RESULT", manifest);
+
+    return {result: {manifest}}
+  }
+
+  function checkIsContractAdmin(idWallet: string) {
+    if (!isContractAdmin(idWallet)) {
+      throw new ContractError("Only admin is allowed to call this function");
+    }
+  }
+
+  function isContractAdmin(idWallet: string) {
+    const wallet = verifyWalletAddress(idWallet);
+    return state.contractAdmins.includes(wallet);
+  }
+
+  function verifyWalletAddress(addr: string): string {
+    const address = addr.toString().trim();
+    if (!/[a-z0-9_-]{43}/i.test(address)) {
+      throw new ContractError('Invalid Arweave address.');
+    }
+
+    return address;
+  }
+
+  function trace(log: string, data: any = {}) {
+    if (state.trace) {
+      console.log(`\n=== ${log} ===\n`, JSON.stringify(data));
+    }
+  }
+
+  function getManifestsForCaller(providerId: string): ManifestData[] {
+    const manifests = providers[providerId].manifests;
+    if (manifests === undefined || manifests.length === 0) {
+      throw new ContractError(`No manifests defined for ${caller}`);
+    }
+    return manifests;
+  }
+
+  function checkPrivileges(caller: string, providerId: string) {
+    if (!providers[providerId].adminsPool.includes(caller)) {
+      throw new ContractError(`${caller} is not an admin for ${providerId}`)
+    }
+  }
+
+  function checkProviderExists(providerId: string) {
+    if (providers[providerId] === undefined) {
+      throw new ContractError(`Provider with id ${providerId} is not registered.`);
+    }
+  }
+
+  function checkProviderId(providerId: string) {
+    if (isEmpty(providerId)) {
+      throw new ContractError("'providerId' field is required.");
+    }
+  }
+
+  function isEmpty(value: string) {
+    return value === undefined || value.length === 0;
+  }
+
```

Comment:
Do we plan to add a function to list all the provider or we're going to get it directly from the state?

Reply:
I was thinking about this, but since "interactRead" SDK function returns current state (and providers can be accessed by `state.providers`), I don't think there is such need...

Reply:
function "providersData" added (with proper status set for all manifests).